cpp/src/arrow/dataset/dataset.h

```diff
@@ -174,9 +186,24 @@ class ARROW_DS_EXPORT Dataset : public std::enable_shared_from_this<Dataset> {
   Dataset(std::shared_ptr<Schema> schema, compute::Expression partition_expression);
 
   virtual Result<FragmentIterator> GetFragmentsImpl(compute::Expression predicate) = 0;
+  virtual Result<FragmentGenerator> GetFragmentsAsyncImpl(compute::Expression predicate);
 
   std::shared_ptr<Schema> schema_;
   compute::Expression partition_expression_ = compute::literal(true);
+
+ private:
```

Comment:
If it's private this means it will not be able to be called by a derived class' `GetFragmentsAsyncImpl`? I'm not sure I understand the respective responsibilities of `GetFragmentsAsyncImpl` vs `GetFragmentsAsyncImplBase` here...

Reply:
I didn't want to introduce an additional argument to the virtual `GetFragmentsAsyncImpl()`, so I moved the parameterized variant of the initial function to a separate one (called `GetFragmentsAsyncImplBase`).

Since this API should be considered experimental, I think we can also change it later if needed.

The reason I didn't make a `virtual GetFragmentsAsyncImpl(..., Executor* = GetCPUThreadPool())` is because in some situations it can cause confusion and probably more serious troubles, please see the link for some common gotchas: http://www.gotw.ca/gotw/005.htm.

I don't see any reason for `GetFragmentsAsyncImplBase` to remain `protected` since it's intended to serve as an impl detail for default base implementation. Derived classes will provide a completely separate impl for `GetFragmentsAsyncImpl`.

Reply:
Default parameters can be confusing in virtual methods, I will agree to that.  Although we have a few places in Arrow where we use them and just follow the maxim of trying to be consistent in the defaults we supply.  With this approach however, I'm confused how a caller would specify which executor to use?  Would it have to be a property of the dataset itself?

Reply:
I think it's ok to add an `Executor*` argument to `GetFragmentsAsyncImpl`. It doesn't need to have a default value, and that sounds better than having 3 methods for the same functionality.